doc/man-source/irpipe.1
.TH IRPIPE "1" "Last change: Aug 2015" "irpipe @version@" "User Commands"
.SH NAME
.P
\fBirpipe\fR - tool to send data to lircd over /dev/irpipe kernel device.
.SH SYNOPSIS
.P
\fBirpipe\fR [\fIoptions\fR] <file
.P
\fBirpipe\fR --read [\fIoptions\fR] >file
.P
\fBirpipe\fR --filter [\fIoptions\fR] <infile >outfile

.SH DESCRIPTION
.P
irpipe is designed to feed data to lircd(8) over a /dev/irpipe device. It's a tool
to enter test data to lircd in a way which is identical to real data from /dev/lirc
devices.
.PP
irpipe uses two formats: text and binary. The binary format is documented in
the lirc(4) manual page. The text format is as defined in mode2(1).
.PP
The irpipe driver allows user  to set the features returned by the
.B LIRC_GET_FEATURES
ioctl. These are documented in lirc(4).
.PP
The irpipe kernel device is often not built in packaged lirc versions. See
[KERNEL DRIVER] below.

.SH OPTIONS
.TP
\fB\-d\fR \fB\-\-device=driver\fR
Set kernel device, default is /dev/irpipe0.
.TP
\fB\-b\fR \fB\-\-bin2text\fR
Convert binary data to text.
.TP
\fB\-t\fR \fB\-\-text2bin\fR
Convert text data to binary.
.TP
\fB\-s\fR \fB\-\-add-sync\fR
Add long initial sync on converted output. This is often required for lircd to
be able to decode data, in particular if keys are not repeated.
.TP
\fB\-f\fR \fB\-\-filter\fR
Send data from stdin to stdout
.TP
\fB\-r\fR \fB\-\-read\fR
Send data from kernel device to stdout. This option is experimental and not
tested.
.TP
\fB\-l\fB\-\-length\fR=\fIbits\fR
Set kernel device signal length. This is only meaningful on LIRCCODE device
emulation.
.TP
\fB\-F\fB\-\-features\fR=\fIbitmask\fR
Set the features bitmask, as returned by the
.B LIRC_GET_FEATURES
ioctl. See lirc(4) for more.
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.TP
\fB\-v\fR \fB\-\-version\fR
Display version.


.SH KERNEL DRIVER

The irpipe tool is designed to work with the irpipe kernel driver, although
the --filter option works without it. This driver is often not built
in packaged versions. If this is the case, the driver can be built from
sources in the drivers/irpipe directory. This requires kernel sources
installed. The drivers/irpipe/README  provides information on building
and installing the driver.

.SH "SEE ALSO"
.B lircd(8)
.br
.B lirc(4)


NEWS
0.9.4

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.
  * The remotes/ directory is not used any more, removed.
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).

0.9.3pre2 1/08/15
  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
0
0.9.3pre1 30/06/15

  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * mode2 has been updated to handle also LIRCCODE devices (#94)
    and the --driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-setup: fixed name clash on recent python versions.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * New build dependency: xsltproc (usually coming with libxml).
  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * Build: now uses parallell make by default.
  * Docs: The Configuration Guide has been updated.

0.9.2 12/09/14
  * Documentation fixes

0.9.2pre3  12/04/14
  * Add fix for #70, by mistake omitted in pre2.

0.9.2pre2  11/30/14
  * Fixes for #67 - #75 besides #71 and #74.

0.9.2pre1  11/21/14
  * Removed functionality:
    + The smode2 program has been retired from the build, although it still
      is available in contrib/.
    + The setup.sh script is removed (build always builds all drivers).
  * Drivers are loaded dynamically. Many programs including lircd now
    needs to know where the drivers are located to be able to use
    them. There are new options, environment variables and hard-coded
    defaults. See the manpages for more.
  * A new tool lirc-lsplugins is added to inspect dynamic driver modules.
  * New, unstable configuration script lirc-setup. The script guides user
    when selecting both capture device and remote config. It sports auto-
    detection of most devices, searching and downloading from the remotes
    database and a considerable amount of bugs.
  * The remotes/ directory has become obsolete, only used for testing.
    New configuration tools uses the remotes database at
    http://sf.net/p/lirc-remotes as the only source (#53). A new tool
    irdb-get(1) supports searching and downloading from remotes database.
  * Building on MacOSX has been refreshed.
  * New devices and features:
    + Support for the Dangerous Prototypes irtoy device is added (#20).
    + Support for for sonyir on MacOSX is added. This is a single hiddev
      device; with the framework in place hopefully more devices will be
      added over time.
    + Support for the DIY assembly kit yard2 usb ir device is added. This is
      an experimental driver which depends on a server which currently is not
      packaged in any major distro; a Debian/Ubuntu PPA is available.
    + Support for the Roku remote protocol added (#8).
    + Support for long keypresses has been added. See the "Configuring LIRC"
      chapter in the manual, search for "ignore_first_event" (#50).
  * lircd has been verified to run as regular user as long as it can open
    devices and such. A new option --effective-user makes lircd drop it's
    privileges after opening devices if started as root. A new udev rule
    is shipped to make /dev/lirc* accessible for users in group 'lirc" (#56).
  * Logging runtime option is renamed: debug -> loglevel. Levels can be given
    symbolically e. g., 'info' or 'error'. Old option name is still supported.
  * Configuration changes
    + The default location for the user lircrc-file has been changed from
      ~/.lircrc to ~/.config/lircrc. The old location is still used as fallback.
      (ticket #52).
    + The lircd.conf configuration has been updated. lircd will now besides
      the lircd.conf file also read all *.conf files in the
      /etc/lirc/lircd.conf.d directory. The already existing include directive
      has been modified to accept wildcards and also been documented.
    + lircd.conf has a new 'driver' attribute for configs requiring a specific

tools/irpipe.cpp
#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>

#include "media/lirc.h"
#include "irpipe.h"


static const char* const help =
	"Write data to irpipe kernel device, applying conversions\n"
	"and ioctls.\n\n"
	"Usage:  irpipe [options] <file\n"
	"        irpipe --read [options] >file\n"
	"        irpipe --filter [options] <infile >outfile\n\n"
	"Options:\n"
	"\t -F --features=bitmask\tSet kernel device features set\n"
	"\t -l --length=bits\tSet kernel device signal length\n"
	"\t -d --device=driver\tSet kernel device, default /dev/irpipe0\n"
	"\t -b --bin2text\t\tConvert binary data to text\n"
	"\t -t --text2bin\t\tConvert text data to binary\n"
	"\t -r --read\t\tSend data from kernel device to stdout\n"
	"\t -f --filter\t\tSend data from stdin to stdout\n"
	"\t -s --add-sync\t\tAdd long initial sync on converted output\n"
	"\t -h --help\t\tDisplay this message\n"
	"\t -v --version\t\tDisplay version\n";


static const struct option options[] = {
	{ "features",	    required_argument, NULL, 'F' },
	{ "length",	    required_argument, NULL, 'l' },
	{ "bin2text",	    no_argument,       NULL, 't' },
	{ "text2bin",	    no_argument,       NULL, 'b' },
	{ "write",	    no_argument,       NULL, 'w' },
	{ "filter",	    no_argument,       NULL, 'f' },
	{ "add-sync",	    no_argument,       NULL, 's' },
	{ "device",	    required_argument, NULL, 'd' },
	{ "help",	    no_argument,       NULL, 'h' },
	{ "version",	    no_argument,       NULL, 'v' },
	{ 0,		    0,		       0,    0	 }
};

static enum {MODE_READ, MODE_WRITE, MODE_FILTER} opt_mode = MODE_WRITE;
static const char* opt_device = "/dev/irpipe0";
static unsigned long opt_features = LONG_MAX;
static int opt_length = -1;
static bool opt_totext = false;
static bool opt_tobin = false;
static bool opt_add_sync = false;


static void parse_options(int argc, char* argv[])
{

	int c;
	long l;
	const char* const optstring = "F:l:btrfd:hsv";

	while (1) {
		c = getopt_long(argc, argv, optstring, options, NULL);
		if (c == -1)
			break;
		switch (c) {
		case 'F':
			l = strtol(optarg, NULL, 10);
			if (l >= INT_MAX || l <= INT_MIN) {
				fprintf(stderr, "Bad numeric: %s\n", optarg);
				exit(1);
			}
			opt_features = (int)l;
		case 'l':
			l = strtol(optarg, NULL, 10);
			if (l >= INT_MAX || l <= INT_MIN) {
				fprintf(stderr, "Bad numeric: %s\n", optarg);
				exit(1);
			}
			opt_length = l;
		case 'b':
			opt_tobin = true;
			break;
		case 's':
			opt_add_sync = true;
			break;
		case 't':
			opt_totext = true;
			break;
		case 'r':
			opt_mode = MODE_READ;
			break;
		case 'f':
			opt_mode = MODE_FILTER;
			break;
		case 'd':
			opt_device = strdup(optarg);
			break;
		case 'h':
			fputs(help, stdout);
			exit(0);
		case 'v':
			fputs("irpipe " VERSION "\n", stdout);
			exit(0);
		default:
			return;
		}
	}
}


/** Redirect newfd to the kernel irpipe device using mode. */
static void irpipe_setup(int newfd, const char* device, int mode)
{
	int fd;

	fd = open(device, mode);
	if (fd == -1) {
		perror("Cannot open kernel device");
		exit(1);
	}
	fd = dup2(fd, newfd);
	if (fd == -1) {
		perror("Cannot dup2 kernel device");
		exit(1);
	}
}


/** Parse a pulse/space text line, return duration. */
static uint32_t process_line(const char* token1, const char* token2)
{
	long value;

	if (token1 == NULL || token2 == NULL)
		return (uint32_t)-1;
	value = strtol(token2, NULL, 10);
	if (value == LONG_MIN || value >= PULSE_BIT || value == 0)
		return (uint32_t)-1;
	if (strcmp("pulse", token1) == 0)
		value |= PULSE_BIT;
	else if (strcmp("space", token1) != 0)
		return (uint32_t)-1;
	return (uint32_t)value;
}


/** Copy stdin to stdout, converting ascii data to binary. */
static void write_tobin(void)
{
	char line[128];
	char buff[128];
	char* token1;
	char* token2;
	int32_t value;

	if (opt_add_sync) {
		value = 1000000;
		if (write(1, &value, sizeof(uint32_t)) != sizeof(uint32_t))
			perror("write() failed.");
	}
	while (fgets(line, sizeof(line), stdin) != NULL) {
		strncpy(buff, line, sizeof(buff) - 1);
		token1 = strtok(buff, "\n ");
		token2 = strtok(NULL, "\n ");
		value = process_line(token1, token2);
		if (value == -1)
			fprintf(stderr,
				"Illegal data (ignored):\"%s\"\n", line);
		else
			if (write(1, &value, sizeof(uint32_t))
			    != sizeof(uint32_t))
				perror("write() failed");
	}
}


/** Copy stdin to stdout, converting binary data to text. */
static void write_totext(void)
{
	char buff[2048 * sizeof(int)];
	int r;
	int i;
	const uint32_t* data;

	if (opt_add_sync)
		puts("space 1000000\n");
	while ((r = read(0, buff, sizeof(buff))) > 0) {
		for (i = 0; i < r; i += sizeof(int)) {
			data = (uint32_t*)(&buff[i]);
			printf("%s %u\n",
			       (*data & PULSE_BIT) ? "pulse" : "space",
			       *data & PULSE_MASK);
		}
	}
	if (r == -1) {
		perror("irpipe: Cannot read stdin");
		exit(1);
	}
}


/** Send verbatim copy of stdin to stdout. */
static void write_raw(void)
{
	char buff[2048 * sizeof(int)];
	int r;
	int written;

	while ((r = read(0, buff, sizeof(buff))) > 0) {
		written = write(1, buff, r);
		if (written != r) {
			perror("irpipe: Cannot write in raw copy");
			exit(1);
		}
	}
	if (r == -1) {
		perror("irpipe: Cannot read stdin");
		exit(1);
	}
}


static void run_ioctl(int cmd, unsigned long arg)
{
	int fd = (opt_mode == MODE_WRITE) ? 1 : 0;
	int r = ioctl(fd, cmd, arg);
	if (r == -1) {
		perror("Cannot run ioctl");
		exit(1);
	}
}


int main(int argc, char**argv)
{
	parse_options(argc, argv);
	if (opt_mode == MODE_WRITE)
		irpipe_setup(1, opt_device, O_WRONLY);
	else if (opt_mode == MODE_READ)
		irpipe_setup(0, opt_device, O_RDONLY);
	if (opt_features != LONG_MAX)
		run_ioctl(LIRC_SET_FEATURES, opt_features);
	if (opt_length != -1)
		run_ioctl(LIRC_SET_LENGTH, opt_length);
	if (opt_tobin)
		write_tobin();
	else if (opt_totext)
		write_totext();
	else
		write_raw();
}
